flink-libraries/flink-gelly/src/main/java/org/apache/flink/graph/library/HITSAlgorithm.java

```diff
@@ -0,0 +1,194 @@
+			this.maxIterations = maxIterations * 2;
+		} else {
+			this.maxIterations = maxIterations * 2 + 1;
+		}
+	}
+
+	/**
+	 * Creates an instance of HITS algorithm.
+	 * If the number of vertices of the input graph is unknown,
+	 * use the {@link HITSAlgorithm#HITSAlgorithm(int, HITSParameter)} constructor instead.
+	 *
+	 * @param maxIterations the maximum number of iterations
+	 * @param hitsParameter the type of final web pages users want to get by this algorithm
+	 */
+	public HITSAlgorithm(int maxIterations, long numberOfVertices, HITSParameter hitsParameter) {
+		if (hitsParameter == HITSParameter.AUTHORITY) {
+			this.maxIterations = maxIterations * 2;
+		} else {
+			this.maxIterations = maxIterations * 2 + 1;
+		}
+		this.numberOfVertices = numberOfVertices;
+	}
+
+	@Override
+	public DataSet<Vertex<K, Double>> run(Graph<K, Double, Double> netGraph) throws Exception {
+		if (this.numberOfVertices == 0) {
+			this.numberOfVertices = netGraph.numberOfVertices();
+		}
+
+		ScatterGatherConfiguration parameter = new ScatterGatherConfiguration();
+		parameter.setDirection(EdgeDirection.ALL);
+		parameter.registerAggregator("sumAllValue", new DoubleSumAggregator());
+
+		return netGraph.runScatterGatherIteration(new VertexUpdate<K>(maxIterations),
+				new MessageUpdate<K>(maxIterations), maxIterations, parameter).getVertices();
+	}
+
+	/**
+	 * Function that updates the value of a vertex by summing up the partial
+	 * values from all messages and normalize the value.
+	 */
+	@SuppressWarnings("serial")
+	public static final class VertexUpdate<K> extends VertexUpdateFunction<K, Double, Double> {
+		private int maxIteration;
+		private DoubleSumAggregator doubleSumAggregator;
+
+		public VertexUpdate(int maxIteration) {
+			this.maxIteration = maxIteration;
+		}
+
+		@Override
+		public void preSuperstep() {
+			doubleSumAggregator = getIterationAggregator("sumAllValue");
+		}
+
+		@Override
+		public void updateVertex(Vertex<K, Double> vertex, MessageIterator<Double> inMessages) {
+			double updateValue = 0;
+
+			for (double element : inMessages) {
+				if (getSuperstepNumber() == maxIteration) {
+					updateValue = element;
+					break;
+				}
+				updateValue += element;
+			}
+
+			if (getSuperstepNumber() != maxIteration) {
+				setNewVertexValue(updateValue);
+				doubleSumAggregator.aggregate(updateValue);
+			} else {
+				setNewVertexValue(vertex.getValue() / updateValue);
+			}
+		}
+	}
+
+	/**
+	 * Distributes the value of a vertex among all neighbor vertices and sum all the
+	 * value in every superstep.
+	 */
+	@SuppressWarnings("serial")
+	public static final class MessageUpdate<K> extends MessagingFunction<K, Double, Double, Double> {
+		private int maxIteration;
+
+		public MessageUpdate(int maxIteration) {
+			this.maxIteration = maxIteration;
+		}
+
+		@Override
+		public void sendMessages(Vertex<K, Double> vertex) {
+			for (Edge<K, Double> edge : getEdges()) {
+				if (getSuperstepNumber() % 2 == 1) {
```

Comment:
The only difference between the if and else is the flipping of source and target, so this could be condensed.